File: examples/python/problem_multicopter_mpc.py
import os
import sys

import crocoddyl
import pinocchio
import numpy as np
import example_robot_data

import multicopter_mpc
from multicopter_mpc.utils.solver_squash import SolverSquashFDDP
import yaml_parser

WITHDISPLAY = 'display' in sys.argv or 'DISPLAY' in os.environ
WITHPLOT = 'plot' in sys.argv or 'PLOT' in os.environ
HECTOR = 'hector' in sys.argv

crocoddyl.switchToNumpyMatrix()

if HECTOR:
    uav = example_robot_data.loadHector()
    link_name = "base_link"
    yaml_uav = yaml_parser.ParserYAML("../../config/multirotor/hector.yaml", "", True)
else:
    uav = example_robot_data.loadIris()
    link_name = "iris__base_link"
    yaml_uav = yaml_parser.ParserYAML("/home/pepms/robotics/libraries/multicopter_mpc/config/multirotor/iris.yaml", "",
                                      True)

uav_model = uav.model
base_link = uav_model.getFrameId(link_name)

# UAV Params
server_uav = yaml_parser.ParamsServer(yaml_uav.getParams())
mc_params = multicopter_mpc.MultiCopterBaseParams()
mc_params.fill(server_uav)

# Mission
yaml_mission = yaml_parser.ParserYAML("/home/pepms/robotics/libraries/multicopter_mpc/config/mission/simple.yaml", "",
                                      True)
server_mission = yaml_parser.ParamsServer(yaml_mission.getParams())
mission = multicopter_mpc.Mission(uav.nq + uav.nv)
mission.fillWaypoints(server_mission)
mission.fillInitialState(server_mission)

dt = 1e-2
state = crocoddyl.StateMultibody(uav_model)
actuation = crocoddyl.ActuationModelMultiCopterBase(state, mc_params.n_rotors, mc_params.tau_f)
problem_mission = multicopter_mpc.ProblemMission(mission, mc_params, uav_model, actuation,
                                                 uav_model.getFrameId(link_name), dt)
problem = problem_mission.createProblem()

ddp_solver = crocoddyl.SolverBoxFDDP(problem)
ddp_solver.setCallbacks([crocoddyl.CallbackLogger(), crocoddyl.CallbackVerbose()])
ddp_solver.solve()

ddp_solver.xs
# if WITHDISPLAY:
#     display = crocoddyl.GepettoDisplay(uav)
#     for idx_wp, wp in enumerate(mission.waypoints):
#         name = 'world/wp' + str(idx_wp)
#         uav.viewer.gui.addXYZaxis(name, [1., 0., 0., 1.], .03, 0.5)
#         wp_pose = pinocchio.SE3ToXYZQUAT(wp.M)
#         uav.viewer.gui.applyConfiguration(name, wp_pose.reshape(7).tolist()[0])

#     display.displayFromSolver(ddp_solver)

File: examples/problem-multicopter-mpc.cpp
#include "pinocchio/parsers/urdf.hpp"
#include "pinocchio/multibody/model.hpp"

#include "example-robot-data/path.hpp"

#include "crocoddyl/core/solvers/ddp.hpp"

#include "yaml_parser/parser_yaml.h"

#include "multicopter_mpc/multicopter-base-params.hpp"
#include "multicopter_mpc/mission.hpp"
#include "multicopter_mpc/problem-mission.hpp"
#include "multicopter_mpc/ocp/trajectory-generator.hpp"

int main(void) {
  pinocchio::Model model;
  pinocchio::urdf::buildModel(EXAMPLE_ROBOT_DATA_MODEL_DIR "/iris_description/robots/iris_simple.urdf",
                              pinocchio::JointModelFreeFlyer(), model);

  yaml_parser::ParserYAML yaml_file("/usr/local/share/multicopter_mpc/multirotor/iris.yaml", "", true);
  yaml_parser::ParamsServer server_params(yaml_file.getParams());

  boost::shared_ptr<multicopter_mpc::MultiCopterBaseParams> mc_params =
      boost::make_shared<multicopter_mpc::MultiCopterBaseParams>();
  mc_params->fill(server_params);

  yaml_parser::ParserYAML yaml_mission("/usr/local/share/multicopter_mpc/mission/passthrough.yaml", "", true);
  yaml_parser::ParamsServer server_mission(yaml_mission.getParams());

  boost::shared_ptr<multicopter_mpc::Mission> mission =
      boost::make_shared<multicopter_mpc::Mission>(model.nq + model.nv);
  mission->fillWaypoints(server_mission);
  mission->fillInitialState(server_mission);

  boost::shared_ptr<pinocchio::Model> mc_model = boost::make_shared<pinocchio::Model>(model);

  multicopter_mpc::TrajectoryGenerator trajectory(mc_model, mc_params,mission, model.getFrameId("iris__base_link"));

  trajectory.createProblem(multicopter_mpc::SolverTypes::BoxFDDP);
  trajectory.setSolverCallbacks(true);
  std::cout << "Here " << std::endl;
  trajectory.solve();
  std::cout << "Here " << std::endl;
  
}